### PR TITLE
[TVOS] - make Kodi core arm64 aware

### PR DESCRIPTION
**Pull request by Memphiz**
This is part 2 of the 3 major pull requests i plan to do for adding TVOS (AppleTV4) support to mainline. 

This basically allows us to compile Kodi core for arm64 targets. I think its pretty straight forward. Once this is in jenkins can build ios builds for arm64 (by simply manually building IOS with arm64 CPU set).

I plan to do beta, rc and final releases for ios 32bit and 64bit aswell once this is in (for Krypton then). Cydia supports providing both builds and doesn't allow to install the 64bit one on 32bit devices so this would be save.

@wsnipex and maybe @koying or whoever is interested in looking it over.

Thx to @MrMC and @linusyang once again for finding the needed bits. @linusyang i think you have a much easier live in the future once this was merged (with your custom 64bit builds hehe).

### CODE REVIEW

**Comment by stefansaraev** [comment]
this also makes #8730 obsolete.

**Comment by wsnipex** [comment]
not really familiar with the apple specific stuff, but went over it and didn't spot anything obvious.

**Comment by koying** [comment]
Same. Seems to be very low impact out of ios. Maybe @popcornmix for info

**Comment by koying** [comment]
@Memphiz half-related: I read that __arm64__ is apple only (vs. __aarch64__). Do you confirm?

**Comment by Memphiz** [comment]
@linusyang or @MrMC for a word? something generic would be good for sure

**Comment by Memphiz** [comment]
maybe i add aarch64 checks in addition for being future proof?

**Comment by MrMC** [comment]
https://www.dropbox.com/s/oewk3j58bo28i3n/clang-arch-arm64.txt?dl=0

that's "clang -dM -arch arm64 -E -x c /dev/null" from a 10.11.2 osx box.

#define __aarch64__ 1
#define __arm64__ 1

so __aarch64__ 'could' be used in place of __arm64__. The only issue would be verifying that any droid arm64 build is ok. 

arm64 is right for target name (appletvos9.1_arm64-target, iphoneos9.2_arm64-target) as this maps into the arch for xcode and that's how the Xcode depends matching magic happens.

@Memphiz looks ok to me.

**Comment by MrMC** [comment]
github ate the leading/trailing underscores :)

**Comment by Memphiz** [comment]
well gcc only defines __arch64__ ... so it would be good to replace all preprocessor checks from __arm64__ to __arch64__ don't you think davilla? (arch is still called arm64 of course - just changing how we check it with the preprocessor i mean)

**Comment by MrMC** [comment]
makes sense but again, since this now hits a flag that linux/droid 'could' generate, someone needs to make sure that other platforms that might be building arm64 are not effected. Or you could take the approach that this is alpha, and anyplace tripped is wrong anyway so just make the change and bang the green button :)

**Comment by Memphiz** [comment]
i think __aarch64__ is there for sure on arm64 toolchains ... __arm64__ only on darwin toolchains ... i have no arm64 toolchain for android at hand for checking but i will change this PR for using the more generic flag :)

**Comment by Memphiz** [comment]
added an overall change to the generic flag on top ...

**Comment by MrMC** [comment]
go for it :) 

Some history, I started using __arm64__ because of sdk, then ran into __aarch64__ with some depends. __arm64__ was so pretty next to __arm__ (symmetry) that I left them this way in core but __aarch64__ is the proper way.

**Comment by Memphiz** [comment]
@popcornmix any arm64 rbpi in the queue you want to check the toolchain for? ^^ :D (just trying to squeeze out some future information - not fully serious here hehe).

**Comment by Memphiz** [comment]
jenkins build this please

**Comment by Memphiz** [comment]
last comment for now - there is no other 64bit arm platform we build for atm afaik - we don't even have a 64bit droid ndk installed on our jenkins slave ... so whoever might hit issues here is someone who cooked it in private ...

**Comment by MrMC** [comment]
nvidia K1 'can be' arm64, nvidia X1 is arm64 all the way I think. might be wrong.

**Comment by Memphiz** [comment]
but we always build arm32bit only atm

File: xbmc/linux/XMemUtils.cpp [hunk 1]
```diff
@@ -95,9 +95,16 @@ void GlobalMemoryStatusEx(LPMEMORYSTATUSEX lpBuffer)
   {
       // Find page size.
       int pageSize;
+#if defined(__aarch64__)
+      // Workaround for 64-bit arm devices (e.g. iPhone 5s, etc.).
+      // sysctl() and host_page_size() give both wrong 4x value (16384).
+      // So, we need to hardcode page size to 4096 here.
+      pageSize = 4096;
+#else
```
Comment (wsnipex) [inline xbmc/linux/XMemUtils.cpp]: this sounds like a darwin only bug?

Comment (MrMC) [inline xbmc/linux/XMemUtils.cpp]: yes, this is a darwin only bug, needs rechecking if iOS/tvOS 9 and above is fixed.

Comment (Memphiz) [inline xbmc/linux/XMemUtils.cpp]: will add a guard with target_darwin then

**Comment by Memphiz** [comment]
added a darwin guard for the pagesize fix and squashed aarch64 into the core commit again - aborted the jenkins build (which was still in the queue anyway) - jenkins build this please

File: xbmc/guilib/GUIFontCache.h [hunk 2]
```diff
@@ -231,10 +231,10 @@ struct CGUIFontCacheDynamicPosition
 
 struct CVertexBuffer
 {
-  void *bufferHandle;
+  unsigned int bufferHandle; // this is really a GLuint
```
Comment (wsnipex) [inline xbmc/guilib/GUIFontCache.h]: this breaks Win32, so needs sync in GUIFontTTFDX.cpp
>(ClCompile target) -> 
         ..\..\xbmc\guilib\GUIFontTTFDX.cpp(194): error C2665: 'CVertexBuffer::CVertexBuffer' : none of the 3 overloads could convert all the argument types [c:\jenkins\slave\workspace\WIN-32\project\VS2010Express\XBMC.vcxproj]
         ..\..\xbmc\guilib\GUIFontTTFDX.cpp(204): error C2446: '!=' : no conversion from 'unsigned int' to 'nullptr' [c:\jenkins\slave\workspace\WIN-32\project\VS2010Express\XBMC.vcxproj]

**Comment by Memphiz** [comment]
should work better now (happend because mrmc has windows removed and i have no windows check it ;) ). jenkins build this please

**Comment by Memphiz** [comment]
if no objectiojs i'll merge this tomorrow

File: addons/library.xbmc.addon/libXBMC_addon.h [hunk 3]
```diff
@@ -47,6 +47,8 @@ typedef intptr_t      ssize_t;
 #define ADDON_HELPER_ARCH       "powerpc-osx"
 #elif defined(__arm__)
 #define ADDON_HELPER_ARCH       "arm-osx"
+#elif defined(__aarch64__)
+#define ADDON_HELPER_ARCH       "arm-osx"
```
Comment (Memphiz) [inline addons/library.xbmc.addon/libXBMC_addon.h]: @Razzeee - that line change is unclear to you - right?

Comment (Razzeee) [inline addons/library.xbmc.addon/libXBMC_addon.h]: Yes, it seems strange to me as above is the same and you could have used an or operator there. Would have expected something like "arm-osx64".
Seems like your grouping stuff and I with my limited knowledge of this part of the code don't understand why you might do that.

Comment (MrMC) [inline addons/library.xbmc.addon/libXBMC_addon.h]: it's a little strange, yes. even stranger is arm-osx to begin with. arm-darwin makes more sense. darwin is the base, then you have darwin/osx, darwin/iOS, darwin/tvOS plus 32 and 64 bit flavors. 

This is all dependent on naming of the actual libs and they must match. 

In reality, one will never have the different arch libs in the sample place so they do not really need different naming.
In practice, it does not matter what naming they have as long as core and libs match.

Comment (Razzeee) [inline addons/library.xbmc.addon/libXBMC_addon.h]: Sure, matching is the important part, but it's super confusing for noobs like me :)

Comment (MrMC) [inline addons/library.xbmc.addon/libXBMC_addon.h]: That's what code comments are for... oh wait, hard core devs never add comment :)

Seriously, I'm a very big fan of commenting concepts and ideas in code for when noobs come along :) Keeps the 'oh this is silly, delete', 'oh snap, what broke ?' action down.

This is an area where I'm still thinking on what to do with it and the side effects.

Comment (Razzeee) [inline addons/library.xbmc.addon/libXBMC_addon.h]: Well that's what I'm trying to say basically , if you think this code is okay like this.

**Comment by MrMC** [comment]
@MrMC makes chicken noises at @Memphiz 
hit the button :)

**Comment by Memphiz** [comment]
@Razzeee @davilla i have cleaned that part up and removed the duplicate irritating definitions and added a comment how the strings have to match ...
https://github.com/Memphiz/xbmc/commit/8c0cd21b0e01b03a9d3143d90d22333a5bf601b2

**Comment by Memphiz** [comment]
jenkins build and merge

**Comment by Razzeee** [comment]
Thank you seems much better to understand

**Comment by Memphiz** [comment]
When reading my own comment it sounds like we could make configure generate those for us @davilla :)

**Pull request merged**